### PR TITLE
General cleanup

### PR DESCRIPTION
**Pull request by acikek**
## Changes

- Updated JDA to `5.0.0-alpha.18`, which introduces heaps of changes, including breaking.
  - Past-deprecated `enable_for`, `disable_for`, `enabled` and `perms` on `discordcommand`.
- Fixed forcing lowercase when creating all application commands
- Deleted `AbstractDiscordCommand` as its `handledError` methods are no longer needed to do async-compatible `Debug` changes in core
- Updated all commands to use the new `autoExecute` syntax
- Added a command type check to the `discord application command` event that utilizes the multi-matching type argument
  - Also on this event: changed the `name` switch to replace spaces with underscores in order to support message and user commands with spaces
- Some general code cleanup where applicable

**NOTE**: At the time of writing, more testing is needed, hence the draft.

### CODE REVIEW

**Comment by acikek** [comment]
Ready for review - [testing script](https://paste.denizenscript.com/View/101093)

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommand.java [hunk 1]
```diff
@@ -92,117 +98,43 @@ public DiscordCommand() {
 
     public enum DiscordInstruction { CONNECT, DISCONNECT, MESSAGE, ADD_ROLE, REMOVE_ROLE, STATUS, RENAME, START_TYPING, STOP_TYPING, EDIT_MESSAGE, DELETE_MESSAGE }
 
-    @Override
-    public void parseArgs(ScriptEntry scriptEntry) throws InvalidArgumentsException {
-        for (Argument arg : scriptEntry) {
-            if (!scriptEntry.hasObject("id")
-                    && arg.matchesPrefix("id")) {
-                scriptEntry.addObject("id", new ElementTag(CoreUtilities.toLowerCase(arg.getValue())));
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommand.java:100]: old code force-lowercased `id` in parseArgs - that needs to be matched in the new autoExecute

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java [hunk 2]
```diff
@@ -1,47 +1,48 @@
...
 import java.util.Map;
 
-public class DiscordCommandCommand extends AbstractDiscordCommand implements Holdable {
+public class DiscordCommandCommand extends AbstractCommand implements Holdable {
 
     public DiscordCommandCommand() {
         setName("discordcommand");
-        setSyntax("discordcommand [id:<id>] [create/perms/delete] (group:<group>) (name:<name>) (type:{slash}/user/message) (description:<description>) (options:<options>) (enabled:{true}/false) (enable_for:<list>) (disable_for:<list>)");
+        setSyntax("discordcommand [id:<id>] [create/delete] (group:<group>) (name:<name>) (type:{slash}/user/message) (description:<description>) (options:<options>)");
         setRequiredArguments(3, 10);
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java]: Maximum edited in meta but not the registration^

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java [hunk 3]
```diff
@@ -177,84 +127,56 @@ public static Command matchCommandByName(ScriptEntry scriptEntry, ElementTag nam
...
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("enable_for") ListTag enableFor,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("disable_for") ListTag disableFor) {
         if (group != null && group.bot == null) {
             group.bot = bot.bot;
         }
         JDA client = bot.getConnection().client;
         Bukkit.getScheduler().runTaskAsynchronously(DenizenDiscordBot.instance, () -> {
             try {
-                if (commandInstruction == null) {
+                Command.Type commandType = type != null ? type : Command.Type.SLASH;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java]: If there should be a default, just set that default in the autoExecute param data

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java [hunk 4]
```diff
@@ -177,84 +127,56 @@ public static Command matchCommandByName(ScriptEntry scriptEntry, ElementTag nam
...
+                if (instruction == null) {
                     Debug.echoError(scriptEntry, "Must have a command instruction!");
                     scriptEntry.setFinished(true);
                     return;
                 }
+                if (enabled != null || enableFor != null || disableFor != null || instruction == DiscordCommandInstruction.PERMS) {
+                    DenizenDiscordBot.oldCommandPermissions.warn(scriptEntry);
+                }
                 else if (name == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java]: this `else` keyword is no longer valid due to the interjection of a separate `if` in the middle

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordConnectCommand.java [hunk 5]
```diff
@@ -201,19 +198,14 @@ public static String flagFilePathFor(String bot) {
...
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("tokenfile") String tokenFile,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("token") SecretTag token,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("intents") ListTag intents) {
         if (tokenFile == null && token == null) {
             throw new InvalidArgumentsRuntimeException("Missing token SecretTag object!");
         }
-        if (scriptEntry.dbCallShouldDebug()) {
-            Debug.report(scriptEntry, getName(), idElement, token, tokenFile, intents);
-        }
-        String id = CoreUtilities.toLowerCase(idElement.asString());
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordConnectCommand.java:216]: why is this deleted?

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCreateChannelCommand.java [hunk 6]
```diff
@@ -71,61 +77,58 @@ public DiscordCreateChannelCommand() {
...
         if (group != null && group.bot == null) {
             group.bot = bot.bot;
         }
+
         Runnable runner = () -> {
             try {
                 ChannelAction<? extends GuildChannel> action;
-                ChannelType typeEnum = type != null ? ChannelType.valueOf(type.asString().toUpperCase()) : ChannelType.TEXT;
-                switch (typeEnum) {
+                ChannelType channelType = type != null ? type : ChannelType.TEXT;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCreateChannelCommand.java]: again specify defaults in the param data

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordInteractionCommand.java [hunk 7]
```diff
@@ -3,36 +3,39 @@
...
 import java.util.List;
 
-public class DiscordInteractionCommand extends AbstractDiscordCommand implements Holdable {
+public class DiscordInteractionCommand extends AbstractCommand implements Holdable {
 
     public DiscordInteractionCommand() {
         setName("discordinteraction");
         setSyntax("discordinteraction [defer/reply/edit/delete] [interaction:<interaction>] (ephemeral) (attach_file_name:<name>) (attach_file_text:<text>) (rows:<rows>) (<message>)");
         setRequiredArguments(2, 7);
         setPrefixesHandled("interaction", "rows", "attach_file_name", "attach_file_text");
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordInteractionCommand.java]: shouldn't have any `setPrefixesHandled` with autoExecute

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordInteractionCommand.java [hunk 8]
```diff
@@ -84,141 +87,104 @@ public DiscordInteractionCommand() {
...
-        }
-        DiscordInteractionInstruction instructionEnum = DiscordInteractionInstruction.valueOf(instruction.asString().toUpperCase());
+    public static void autoExecute(ScriptEntry scriptEntry,
+                                   @ArgName("instruction") DiscordInteractionInstruction instruction,
+                                   @ArgPrefixed @ArgName("interaction") DiscordInteractionTag interaction,
+                                   @ArgName("ephemeral") boolean ephemeral,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("attach_file_name") String attachFileName,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("attach_file_text") String attachFileText,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("rows") ObjectTag rows,
+                                   @ArgLinear @ArgDefaultNull @ArgName("message") ObjectTag message) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordInteractionCommand.java]: message needs `@ArgRaw`, as you can see the `getRawObject` in the original parseArgs

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordInteractionCommand.java [hunk 9]
```diff
@@ -84,141 +87,104 @@ public DiscordInteractionCommand() {
...
-    public void execute(ScriptEntry scriptEntry) {
-        ElementTag instruction = scriptEntry.getElement("instruction");
-        DiscordInteractionTag interaction = scriptEntry.requiredArgForPrefix("interaction", DiscordInteractionTag.class);
-        boolean ephemeral = scriptEntry.argAsBoolean("ephemeral");
-        ElementTag attachFileName = scriptEntry.argForPrefixAsElement("attach_file_name", null);
-        ElementTag attachFileText = scriptEntry.argForPrefixAsElement("attach_file_text", null);
-        ObjectTag rows = scriptEntry.argForPrefix("rows", ObjectTag.class, true);
-        ObjectTag message = scriptEntry.getObjectTag("message");
-        if (scriptEntry.dbCallShouldDebug()) {
-            // Note: attachFileText intentionally at end
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordInteractionCommand.java:113]: `// Note: attachFileText intentionally at end` this comment needs to be reinstated

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommand.java [hunk 10]
```diff
@@ -92,117 +98,44 @@ public DiscordCommand() {
...
-        if (scriptEntry.dbCallShouldDebug()) {
-            Debug.report(scriptEntry, getName(), id, channel, instruction, message, user, guild, role, status, activity, url, tokenFile, messageId);
-        }
+    public static void autoExecute(ScriptEntry scriptEntry,
+                                   @ArgPrefixed @ArgName("id") String idString,
+                                   @ArgName("instruction") DiscordInstruction instruction,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("code") String code,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("tokenfile") String tokenFile,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("channel") DiscordChannelTag channel,
+                                   @ArgLinear @ArgDefaultNull @ArgName("message") String message,
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommand.java:111]: this needs to be marked as Raw

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java [hunk 11]
```diff
@@ -177,84 +127,55 @@ public static Command matchCommandByName(ScriptEntry scriptEntry, ElementTag nam
...
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("enable_for") ListTag enableFor,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("disable_for") ListTag disableFor) {
         if (group != null && group.bot == null) {
             group.bot = bot.bot;
         }
         JDA client = bot.getConnection().client;
         Bukkit.getScheduler().runTaskAsynchronously(DenizenDiscordBot.instance, () -> {
             try {
-                if (commandInstruction == null) {
+                if (instruction == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java:148]: This null check is redundant

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java [hunk 12]
```diff
@@ -177,84 +127,55 @@ public static Command matchCommandByName(ScriptEntry scriptEntry, ElementTag nam
...
                     Debug.echoError(scriptEntry, "Must have a command instruction!");
                     scriptEntry.setFinished(true);
                     return;
                 }
-                else if (name == null) {
+                if (enabled != null || enableFor != null || disableFor != null || instruction == DiscordCommandInstruction.PERMS) {
+                    DenizenDiscordBot.oldCommandPermissions.warn(scriptEntry);
+                }
+                if (name == null) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCommandCommand.java:156]: also redundant

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCreateChannelCommand.java [hunk 13]
```diff
@@ -71,61 +77,57 @@ public DiscordCreateChannelCommand() {
...
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("description") String description,
+                                   @ArgPrefixed @ArgDefaultText("text") @ArgName("type") ChannelType type,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("category") String category,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("position") ElementTag position,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("roles") ListTag roles,
+                                   @ArgPrefixed @ArgDefaultNull @ArgName("roles") ListTag users) {
         if (group != null && group.bot == null) {
             group.bot = bot.bot;
         }
+
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordCreateChannelCommand.java:93]: blank line shouldn't be here

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordModalCommand.java [hunk 14]
```diff
@@ -64,39 +68,34 @@ public DiscordModalCommand() {
...
                 if (!interaction.interaction.isAcknowledged()) {
                     IModalCallback replyTo = (IModalCallback) interaction.interaction;
-                    Modal modal = Modal.create(name.toString(), title.toString()).addActionRows(actionRows).build();
+                    Modal modal = Modal.create(name, title).addActionRows(actionRows).build();
                     ModalCallbackAction action = replyTo.replyModal(modal);
                     action.complete();
                 }
                 else {
-                    handleError(scriptEntry, "Interaction already acknowledged!");
-                    return;
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordModalCommand.java:95]: this `return` shouldn't have been deleted

File: src/main/java/com/denizenscript/ddiscordbot/commands/DiscordReactCommand.java [hunk 15]
```diff
@@ -172,45 +141,31 @@ public void execute(ScriptEntry scriptEntry) {
...
-                    }
-                    else {
-                        action = msg.clearReactions(reaction.asString());
-                    }
+                    action = msg.clearReactions(emoji);
                 }
                 break;
             }
-            default: {
-                return; // Not possible, but required to prevent compiler error
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/commands/DiscordReactCommand.java:207]: as noted on Discord this was actually intended to be there

File: src/main/java/com/denizenscript/ddiscordbot/events/DiscordApplicationCommandScriptEvent.java [hunk 16]
```diff
@@ -60,7 +62,11 @@ public boolean matches(ScriptPath path) {
         if (!tryGuild(path, getEvent().isFromGuild() ? getEvent().getGuild() : null)) {
             return false;
         }
-        if (!runGenericSwitchCheck(path, "name", getEvent().getName())) {
+        String type = path.eventArgLowerAt(1);
+        if (!type.equals("application") && !CoreUtilities.equalsIgnoreCase(type, getEvent().getCommandType().name())) {
```
Comment (mcmonkey4eva) [inline src/main/java/com/denizenscript/ddiscordbot/events/DiscordApplicationCommandScriptEvent.java:66]: should be a generic check not equality
